rs/state_manager/tests/state_manager.rs

```diff
@@ -1579,6 +1579,78 @@ fn can_remove_checkpoints_and_noncheckpoints_separately() {
     });
 }
 
+#[test]
+fn remove_inmemory_states_below_can_keep_extra_states() {
+    state_manager_restart_test(|state_manager, restart_fn| {
+        let mut heights = vec![height(0)];
+        for i in 1..10 {
+            let (_height, state) = state_manager.take_tip();
+            heights.push(height(i));
+
+            let scope = if i % 2 == 0 {
+                CertificationScope::Full
+            } else {
+                CertificationScope::Metadata
+            };
+
+            state_manager.commit_and_certify(state, height(i), scope.clone(), None);
+        }
+        // We need to wait for hashing to complete, otherwise the
+        // checkpoint can be retained until the hashing is complete.
+        state_manager.flush_tip_channel();
+
+        assert_eq!(state_manager.list_state_heights(CERT_ANY), heights);
+
+        state_manager.remove_inmemory_states_below(height(9), &btreeset![height(7), height(8)]);
```

Comment:
It would be good to test some edge cases.
For example, if `extra_states_to_keep` wasn't right below the cutoff height. Maybe also include `height(3)`? Similarly a height greater than 9 might also be another edge case we might want to test.